src/mbgl/style/expression/within.cpp

```diff
@@ -0,0 +1,173 @@
+#include <mbgl/style/expression/within.hpp>
+
+#include <mapbox/geojson.hpp>
+#include <mapbox/geometry.hpp>
+#include <mbgl/style/conversion_impl.hpp>
+#include <mbgl/tile/geometry_tile_data.hpp>
+#include <mbgl/util/logging.hpp>
+#include <mbgl/util/string.hpp>
+
+#include <rapidjson/document.h>
+#include <mbgl/style/conversion/json.hpp>
+
+namespace mbgl {
+namespace {
+
+double isLeft(mbgl::Point<double> P0, mbgl::Point<double> P1, mbgl::Point<double> P2) {
+    return ((P1.x - P0.x) * (P2.y - P0.y) - (P2.x - P0.x) * (P1.y - P0.y));
+}
+
+// winding number algorithm for checking if point inside a ploygon or not.
+// http://geomalgorithms.com/a03-_inclusion.html#wn_PnPoly()
+bool pointWithinPolygons(mbgl::Point<double> point, const mapbox::geometry::polygon<double>& polys) {
+    //  wn = the winding number (=0 only when point is outside)
+    int wn = 0;
+    for (auto poly : polys) {
+        auto size = poly.size();
+        // loop through every edge of the polygon
+        for (decltype(size) i = 0; i < size - 1; ++i) {
+            if (poly[i].y <= point.y) {
+                if (poly[i + 1].y > point.y) { // upward horizontal crossing from point to edge E(poly[i], poly[i+1])
+                    if (isLeft(poly[i], poly[i + 1], point) > 0) {
+                        ++wn;
+                    }
+                }
+            } else {
+                if (poly[i + 1].y <= point.y) { // downward crossing
+                    if (isLeft(poly[i], poly[i + 1], point) < 0) {
+                        --wn;
+                    }
+                }
+            }
+        }
+        if (wn != 0) {
+            return true;
+        }
```

Comment:
Here, you return a result immediately if you're inside the first ring, even if subsequent holes negate that result — this leads to incorrect results with holes.

Another small gotcha is that GL Native resolves overlapping rings in a polygon [with an even-odd rule](https://github.com/mapbox/mapbox-gl-native/blob/f53143f1a1018050a2104f9afd929c963951527d/src/mbgl/tile/geometry_tile_data.cpp#L49) instead of a non-zero one, so you want to return `true` if `wn` is even, not necessarily zero.

If so, you could remove the branching into two loops and simplify the code — here's an example from one of my library of the same routine: https://github.com/mapbox/which-polygon/blob/master/index.js#L81-L95

Reply:
One more point — the naming `pointWithinPolygons` is a little misleading, since you're testing against a _single_ polygon which has multiple _rings_ (an outer ring and holes), not multiple polygons.

Reply:
fixed!